Bats readme (#4328)
rancher-sandbox/rancher-desktop: bats/README.md

## Overview

BATS is a testing framework for Bash shell scripts that provides supporting libraries and helpers for customizable server-side test automation.

## Setup

Run the following commands from the root directory of the Git repository to install BATS and its helper libraries into the BATS test directory.

```
git submodule update --init
```

## Running BATS

1. To run the BATS test, specify the path to BATS executable from bats-core and run the following commands:

    To run a specific test set from a bats file:

      Example:

      ```
      cd bats
      ./bats-core/bin/bats tests/registry/creds.bats
      ```

    To run all BATS tests:

      ```
      cd bats
      ./bats-core/bin/bats make all
      ```

    To run the BATS test, specifying some of Rancher Desktop's configuration, run the following commands:

      Example:

      ```
      cd bats
      RD_CONTAINER_RUNTIME=moby RD_USE_IMAGE_ALLOW_LIST=false ./bats-core/bin/bats tests/registry/creds.bats
      ```

## Writing BATS Tests

1. Add BATS test by creating files with `.bats` extension under ./bats/tests/FOLDER_NAME

> 2. A Bats test file is a Bash script with special syntax for defining test cases. Under the hood, each test case is just a function with a description.

Author: do you want me to remove it or to keep this as a simple brief intro the way it is?

> 2. A Bats test file is a Bash script with special syntax for defining test cases. Under the hood, each test case is just a function with a description.

Reviewer: I would remove it unless you have some good argument why you want to keep it. It is not self-explanatory, so you still need to read the docs anyways. And we have plenty of tests that could work as an example.

Maybe have one example extensively documented, and then link to the actual test from here as a "recommended tutorial"? But maybe this should wait until later. Just get a README published, and then refine over time.

> 2. A Bats test file is a Bash script with special syntax for defining test cases. Under the hood, each test case is just a function with a description.

Author: okay, I deleted the existing example and kept only the links to the documentation. I can elaborate a tutorial in a follow up PR

          ```
          sudo apt install shellcheck
          ```

        - `Shfmt` is available as snap application. If your distribution has snap installed, you can install shfmt using command:

          ```
          sudo snap install shfmt
          ```
          The other way to install Shfmt is by using the following one-liner command:

          ```
          curl -sS https://webinstall.dev/shfmt | bash
          ```
          If you have Go installed, you can also install `shfmt` by running:

          ```
          go install mvdan.cc/sh/v3/cmd/shfmt@v3.6.0
          ```
    On Windows:

        - The simplest way to install ShellCheck locally is:

          Via chocolatey:

            ```
            choco install shellcheck
            ```
          Via scoop:

            ```
            `scoop install shellcheck`
            ```
        - If you have Go installed, you can install `shfmt` by running:

          ```
          go install mvdan.cc/sh/v3/cmd/shfmt@v3.6.0
          ```

  2. Get the syntax and formatting feedback for BATS linting by running:

      ```
      make -C bats lint
      ```

> 3. Please, make sure to fix the highlighted linting errors prior to submitting a PR. Not required, but highly recommendable. To do so, run the following command:

Reviewer: Fixing linter errors is **required** now; Github actions will fail if any errors are detected.

```suggestion
3. Please, make sure to fix the highlighted linting errors prior to submitting a PR. You can automatically apply formatting changes suggested by `shfmt` by running the following command:
```

## Overview

> BATS is a testing framework for Bash shell scripts that provides supporting libraries and helpers for customizable server-side test automation.

Reviewer: Not sure what "server-side" means here; Rancher Desktop is a desktop application.

```suggestion
BATS is a testing framework for Bash shell scripts that provides supporting libraries and helpers for customizable test automation.
```

## Setup

> It's important to have a Rancher Desktop CI build installed with privileged rights before running the BATS tests.

Reviewer: Let's remove "with privileged rights" here, assuming it just means "installed for all users". I'll create a PR to fall back on a per-user installation if the system-wide location does not exist.

```suggestion
It's important to have a Rancher Desktop CI or release build installed before running the BATS tests.
```

On Windows:
  Prior to running `git submodule update --init`, it's important to setup the Git configuration by running the following commands:

> 

Reviewer: I think you should explain that you need to clone the repo first.

Note that changing crlf settings is not necessary when you clone inside a WSL distro; it is only needed when you clone on Win32.

  ```
  git config --global core.eol lf
  git config --global core.autocrlf false
  ```
  Then, checkout the Git repository or clone it.

All platforms:
From the root directory of the Git repository, run the following commands to install BATS and its helper libraries into the BATS test directory:

  ```
  git submodule update --init
  ```

## Running BATS

> 1. To run the BATS test, specify the path to BATS executable from bats-core and run the following commands:

Reviewer: Sine you talk above about cloning the repo on Windows, you should also explain how to locate it inside a WSL distro on `/mnt/...`.

I see that happens in the 3rd example, but it does not explain that `/mnt/c` represents the `C:` drive from windows.
